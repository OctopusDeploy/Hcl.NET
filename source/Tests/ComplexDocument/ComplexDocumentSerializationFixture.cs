using System;
using System.Collections.Generic;
using Assent;
using NUnit.Framework;
using Octopus.Ocl;

namespace Tests.ComplexDocument
{
    public class ComplexDocumentSerializationFixture
    {
        DeploymentProcess GetTestData()
            => new DeploymentProcess
            {
                Steps = new List<DeploymentStep>
                {
                    new DeploymentStep("Simple Script Step")
                    {
                        Roles = new List<string>
                            { "web-server" },
                        Actions = new List<DeploymentAction>
                        {
                            new DeploymentAction
                            {
                                Name = "Simple Script Action",
                                Type = "Inline Script Action",
                                Properties = new Dictionary<string, string>
                                {
                                    { "Syntax", "PowerShell" },
                                    { "Body", "Write-Host 'Hi'" }
                                }
                            }
                        }
                    },
                    new DeploymentStep("Rolling Step")
                    {
                        Roles = new List<string>
                            { "role1", "role2" },
                        Actions = new List<DeploymentAction>
                        {
                            new DeploymentAction
                            {
                                Name = "Deploy Website",
                                Type = "Deploy to IIS",
                                Properties = new Dictionary<string, string>
                                {
                                    { "AppPool.Framework", "v4.0" },
                                    { "AppPool.Identity", "ApplicationPoolIdentity" },
                                    { "CustomField", "Another value" }
                                }
                            },
                            new DeploymentAction
                            {
                                Name = "Deploy Website",
                                Type = "Deploy to IIS",
                                Properties = new Dictionary<string, string>
                                {
                                    { "AppPool.Framework", "v4.0" },
                                    { "AppPool.Identity", "ApplicationPoolIdentity" },
                                    { "CustomField", "Another value" }
                                }
                            }
                        }
                    }
                }
            };

        [Test]
        public void ToOclDocument()
        {
            var options = new OclSerializerOptions
            {
                Converters = new List<IOclConverter>
                {
                    new DeploymentActionOclConverter()
                }
            };

            var obj = new DeploymentStep("MyStep")
            {
                Actions = new List<DeploymentAction>
                {
                    new DeploymentAction { Type = "Decisive", Name = "Run" },
                    new DeploymentAction { Type = "Take No", Name = "Sit" }
                }
            };

            var result = OclConvert.ToOclDocument(obj, options);

            result.Should()
                .Be(
                    new OclDocument
                    {
                        new OclAttribute("name", "MyStep"),
                        new OclBlock("decisive", new[] { "run" }),
                        new OclBlock("take_no", new[] { "sit" })
                    }
                );
        }

        [Test]
        public void Serialization()
        {
            var options = new OclSerializerOptions
            {
                Converters = new List<IOclConverter>
                {
                    new DeploymentStepOclConverter(),
                    new DeploymentActionOclConverter()
                }
            };

            var result = OclConvert.Serialize(GetTestData(), options);
            this.Assent(result);
        }
    }
}