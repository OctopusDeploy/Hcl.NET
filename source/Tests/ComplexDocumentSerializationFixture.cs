using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using FluentAssertions;
using NUnit.Framework;
using Octopus.Hcl;
using Octopus.Hcl.Converters;

namespace Tests
{
    public class ComplexDocumentSerializationFixture
    {
        [Test]
        public void ToHDocument()
        {
            var options = new HclSerializerOptions()
            {
                Converters = new List<IHclConverter>()
                {
                    new ActionConverter()
                }
            };

            var obj = new ThoughtProcess()
            {
                Actions = new List<Action>()
                {
                    new Action { Type = "Decisive", Name = "Run" },
                    new Action { Type = "Take No", Name = "Sit" }
                }
            };
            HclConvert.ToHDocument(obj, options)
                .Should()
                .BeEquivalentTo(
                    new HDocument()
                    {
                        new HBlock("decisive", new[] { "run" }),
                        new HBlock("take_no", new[] { "sit" })
                    }
                );
        }

        class ActionConverter : BlockHclConverter
        {
            public override bool CanConvert(Type type)
                => typeof(Action).IsAssignableFrom(type);

            protected override string GetName(string name, object obj)
                => ((Action)obj).Type.Replace(" ", "_").ToLower();

            protected override IEnumerable<string> GetLabels(object obj)
                => new[] { ((Action)obj).Name.ToLower() };

            protected override IEnumerable<IHElement> GetElements(object obj, HclConversionContext context)
                => GetElements(
                    obj,
                    from p in obj.GetType().GetProperties()
                    where p.CanRead
                    where p.Name != nameof(Action.Type)
                    where p.Name != nameof(Action.Name)
                    select p,
                    context
                );
        }

        class ThoughtProcess
        {
            public List<Action> Actions { get; set; } = new List<Action>();
        }

        class Action
        {
            public string Type { get; set; } = "";
            public string Name { get; set; } = "";
        }
    }
}