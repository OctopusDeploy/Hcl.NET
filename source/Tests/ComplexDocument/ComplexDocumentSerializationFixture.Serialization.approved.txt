inline_script_action "Simple Script Action" {
    roles = ["web-server"]
    syntax = "PowerShell"
    body = "Write-Host 'Hi'"
}

rolling "Rolling Step" {
    roles = ["role1", "role2"]

    deploy_to_iis "Deploy Website" {
        apppool_framework = "v4.0"
        apppool_identity = "ApplicationPoolIdentity"
        customfield = "Another value"
    }

    deploy_to_iis "Deploy Website" {
        apppool_framework = "v4.0"
        apppool_identity = "ApplicationPoolIdentity"
        customfield = "Another value"
    }
}